### PR TITLE
FEAT: Add Tally Ho Wallet Package

### PR DESCRIPTION
**Pull request by 0xzoz**
[Tally Ho](https://tally.cash/) is an open-source community-run wallet incubated by the team at [Thesis](https://thesis.co/) and community contributors.

As we near the release of the DAO, It is essential to have the ability to put the brand front and center in applications within the ecosystem. We have had numerous projects using the awesome Rainbow library reach out about using Tally Ho not as a standard injected provider.

This feature adds a Tally Ho connector to the library and sets it as a default wallet. 

NOTE: Due to rainbow utilizing the wagmi framework there are some extra files to make the adapter compatible eg `types.ts` & `tallyHoConnector.ts` . There has been work done in the wagmi library [here](https://github.com/wagmi-dev/wagmi/pull/1055) and a discussion regarding a partnership to add it. Once this is completed I will be able to remove the code and it will reduce the footprint in Rainbow.

### CODE REVIEW

**Comment by Naxsun** [comment]
Thanks for creating this PR zoz! Hope this PR gets approved to increase our dApp connection coverage and UX!

**Comment by mhluongo** [comment]
Thanks for tackling this @0xzoz! RainbowKit team, let us know if there's anything we can do to help support this on the team or community side!

**Comment by JVK-6935** [comment]
+1

**Comment by am1993421** [comment]
+1

**Comment by 0xDaedalus** [comment]
Awesome PR @0xzoz - really looking forward to this 😀

**Comment by 0xzoz** [comment]
@Tesbarmix1 Do you need me to resolve these conflicts? Or is there anything you need from my end?

**Comment by nickbytes** [comment]
@0xzoz – could you please resolve these conflicts?

**Comment by 0xzoz** [comment]
> @0xzoz – could you please resolve these conflicts? Once done, I can look into getting this merged.

Thanks @nickbytes. Should be good to go.

**Comment by nickbytes** [comment]
@0xzoz seems like something in this PR is causing RainbowKit not to recognize Metamask is installed when Tally Ho is enabled. I've tested it in the deploy preview here: https://rainbowkit-example-rb4nmlz8a-rainbowdotme.vercel.app/

File: packages/rainbowkit/src/wallets/getDefaultWallets.ts
Comment (nickbytes) [inline packages/rainbowkit/src/wallets/getDefaultWallets.ts:28]: We're not going to add anymore wallets to `getDefaultWallets`. This was only meant to be used internally for demos and testing. Moving forward, we'd like the community to publish their own versions of this helper function and include whatever wallets they find best for their users.

**Review by nickbytes — CHANGES_REQUESTED** [review]
Need to resolve Metamask/Tally Ho compatibility issues + remove from `getDefaultWallets` before merging.

**Comment by nickbytes** [comment]
@0xzoz what's status of Tally Ho connector being added to wagmi?

**Comment by 0xzoz** [comment]
> https://rainbowkit-example-rb4nmlz8a-rainbowdotme.vercel.app/

@nickbytes Do you have the Tally Ho wallet setting 'Use Tally Ho as your default wallet' on? Tally Ho has a feature that will hijack window and override Metamask.

With this unset it works as desired

Screen Shot 2022-10-31 at 7 50 20 PM

File: packages/rainbowkit/src/wallets/getDefaultWallets.ts
Comment (0xzoz) [reply packages/rainbowkit/src/wallets/getDefaultWallets.ts]: Makes sense. Removed

**Comment by 0xzoz** [comment]
> @0xzoz what's status of Tally Ho connector being added to wagmi?

@nickbytes it is still in discussions. I will follow up today as that will be a huge win for us. If we could get support from rainbow to suggest it or chime in on the PR it would be very much appreciated. No stress if not though.

> Need to resolve Metamask/Tally Ho compatibility issues + remove from getDefaultWallets before merging.

Default wallets is addressed and having Tally not set as default should address the other issue.

**Comment by 0xzoz** [comment]
> Thanks for the updates @0xzoz – let me know how convos w/ wagmi go

No problem. Will it be able to be merged before this or will it require the wagmi integration first? I'm more than happy to double back and adjust once it is integrated with wagmi if it is merged prior.

**Comment by 0xzoz** [comment]
@nickbytes we had a call with Tom from Wagmi late last week. They are looking at a more sustainable way to add connectors/wallets. This is likely not in the coming weeks. The best option in the near term is to have this PR merged and then revisit it once wagmi has completed the design. This will then allow us to remove 95% of the code in this PR(which I will happily take care of).

Let me know your thoughts

**Comment by 0xzoz** [comment]
@nickbytes any thoughts on the above?

**Comment by nickbytes** [comment]
> > https://rainbowkit-example-rb4nmlz8a-rainbowdotme.vercel.app/
> 
> @nickbytes Do you have the Tally Ho wallet setting 'Use Tally Ho as your default wallet' on? Tally Ho has a feature that will hijack window and override Metamask.
> 
> With this unset it works as desired
> 
> Screen Shot 2022-10-31 at 7 50 20 PM

@0xzoz this is the default behavior for Tally Ho?

**Comment by nickbytes** [comment]
> @nickbytes we had a call with Tom from Wagmi late last week. They are looking at a more sustainable way to add connectors/wallets. This is likely not in the coming weeks. The best option in the near term is to have this PR merged and then revisit it once wagmi has completed the design. This will then allow us to remove 95% of the code in this PR(which I will happily take care of).
> 
> Let me know your thoughts

I don't like the idea that we are going to be putting something in RainbowKit that is actually a feature of wagmi, but it looks like there's no way around this presently if we'd like to get something to work for Tally Ho. I understand that wagmi is in the same predicament that RainbowKit is: in order to support every wallet, you must take on the burden of maintaining every wallet's connector.

A bit of a sidenote, but I've been wondering if a more sustainable way to do this may be to have team's publish their own wallet/connector packages, and simply link to them on the RainbowKit github page. Teams would then be responsible for updating their own wallet/connectors as wagmi and RainbowKit update, rather than the RainbowKit team having to maintain these.

**Comment by 0xzoz** [comment]
> @0xzoz this is the default behavior for Tally Ho? This basically makes other injected wallets unusable if a dapp enabled Tally Ho. This seems like a confusing experience for developers and users. Is there any way we can fix this? Or am I overthinking this?

This was intentional and as a remedy for the poor UX experience initially due to MM setting the wallet standard by default. The idea was that if a user had Tally Ho installed they could set this and not have issues with conflicting browser wallets. Tally also injects the tally object in window which is what we will be using in this integration. I think this feature will be sunsetted once the industry develops better wallet standards so wallets like Tally Ho and the new Rainbow browser wallet(congrats! excited to see it) can work harmoniously together.

**Comment by 0xzoz** [comment]
> I don't like the idea that we are going to be putting something in RainbowKit that is actually a feature of wagmi, but it looks like there's no way around this presently if we'd like to get something to work for Tally Ho. I understand that wagmi is in the same predicament that RainbowKit is: in order to support every wallet, you must take on the burden of maintaining every wallet's connector.

We totally agree! I will personally make it my job to be at your disposal for this initial lift while we go through this transition period. The wagmi team has made some changes in the most recent release(0.9.0) that will allow me to remove the majority of the code in this PR. I went to make the changes but will have to wait until RainbowKit bumps the package. For context [here are the changes made](https://github.com/wagmi-dev/wagmi/pull/1406) and [documentation](https://wagmi.sh/react/connectors/injected#getprovider)

> A bit of a sidenote, but I've been wondering if a more sustainable way to do this may be to have team's publish their own wallet/connector packages, and simply link to them on the RainbowKit github page. Teams would then be responsible for updating their own wallet/connectors as wagmi and RainbowKit update, rather than the RainbowKit team having to maintain these.

I think this is a great idea. I personally have not thought deeply enough on this but Tom from wagmi has and did mention exactly the same thoughts. It would be awesome to develop and set a standard.


Apologies for my late reply. I saw this message late last week and was afk.

**Comment by 0xzoz** [comment]
@nickbytes I have removed the redundant code and refactored as per Wagmi changes to injected connector in v0.9.6

**Comment by DanielSinclair** [comment]
Thanks for making the changes and working on getting the injector improvements into Wagmi @0xzoz. We'll try to review this soon and get it into our next batch of wallet connector additions.

File: packages/rainbowkit/src/wallets/walletConnectors/tallyHoWallet/tallyHoWallet.ts
Comment (DanielSinclair) [inline packages/rainbowkit/src/wallets/walletConnectors/tallyHoWallet/tallyHoWallet.ts:44]: Installed should have logic to detect if the provider is injected, or not. Reference the MetaMask example: https://github.com/rainbow-me/rainbowkit/blob/78f6cbea363e89ceffa4132b084f14ae90bbf825/packages/rainbowkit/src/wallets/walletConnectors/metaMaskWallet/metaMaskWallet.ts#L56

**Review by DanielSinclair — CHANGES_REQUESTED** [review]

**Comment by DanielSinclair** [comment]
Hey @0xzoz! Can you turn on [Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests). Going to pull in a few tweaks from [@daniel/tally-ho-wallet](https://github.com/rainbow-me/rainbowkit/tree/%40daniel/tally-ho-wallet)

**Comment by 0xzoz** [comment]
> Hey @0xzoz! Can you turn on [Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)? Going to pull in a few tweaks from [@daniel/tally-ho-wallet](https://github.com/rainbow-me/rainbowkit/tree/%40daniel/tally-ho-wallet)

Yep, Its on.

File: packages/rainbowkit/src/wallets/walletConnectors/tallyHoWallet/tallyHoWallet.ts
Comment (DanielSinclair) [reply packages/rainbowkit/src/wallets/walletConnectors/tallyHoWallet/tallyHoWallet.ts:44]: Bumping this; last implementation issue

**Review by DanielSinclair — CHANGES_REQUESTED** [review]
Upon testing, am bumping into unexpected behavior where the Tally Ho extension launches twice on dApp load (2 new windows). Replicating the MetaMask behavior is preferred (i.e. user initiates Connect). Can this be addressed prior to release?
Screenshot 2023-02-07 at 2 23 53 AM

**Comment by DanielSinclair** [comment]
Apologies for opening and closing. There appears to be a Vercel bug blocking our `rainbowkit-example` deployment. We'll proceed reviewing either way.

File: packages/rainbowkit/src/wallets/walletConnectors/tallyHoWallet/tallyHoWallet.ts
Comment (0xzoz) [reply packages/rainbowkit/src/wallets/walletConnectors/tallyHoWallet/tallyHoWallet.ts]: addressed this in the latest commit

**Comment by 0xzoz** [comment]
> Upon testing, am bumping into unexpected behavior where the Tally Ho extension launches twice on dApp load (2 new windows). Replicating the MetaMask behavior is preferred (i.e. user initiates Connect). Can this be addressed prior to release? Screenshot 2023-02-07 at 2 23 53 AM

I believe this recently merged fix addresses the issue. It will be released shortly. I will reach out once Ive tested and confirmed the fix works

**Comment by 0xzoz** [comment]
@DanielSinclair I just tested the 22.01 Tally Ho release and the issue is fixed. We should be good to get this into prod 🐶

**Comment by 0xzoz** [comment]
Just checking in to see if there is anything you need from us @DanielSinclair ?

**Comment by DanielSinclair** [comment]
@0xzoz Seeing a deployment abd check error
```bash
packages/example build: node:internal/process/promises:288
packages/example build:             triggerUncaughtException(err, true /* fromPromise */);
packages/example build:             ^
packages/example build: ReferenceError: window is not defined
packages/example build:     at tallyHoWallet (/home/runner/work/rainbowkit/rainbowkit/packages/example/.next/server/pages/_app.js:1520:22)
packages/example build:     at /home/runner/work/rainbowkit/rainbowkit/packages/example/.next/server/pages/_app.js:129:100 {
packages/example build:   type: 'ReferenceError'
packages/example build: }
packages/example build: Node.js v18.14.0
packages/example build: Failed
undefined
/home/runner/work/rainbowkit/rainbowkit/packages/example:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  example@0.0.42 build: `next build`
Exit status 1
 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by 0xzoz** [comment]
@DanielSinclair sorry I missed this last comment. I have added a new PR that fixes this and that also addresses the rebrand from Tally Ho to Taho.

See #1077

**Pull request closed without merging**
